db-testsuite/src/match.ts

```diff
@@ -0,0 +1,248 @@
+test('lt checks for same type', valueDoesntMatch, 3, 'lt', '4');
+test('lt checks for same type', keyMatches, 'abc', 'lt', 'abd');
+test('lt checks for same type', keyDoesntMatch, 'abc', 'lt', 4 as any);
+
+test('gte checks for same type', valueMatches, 3, 'gte', 2);
+test('gte checks for same type', valueMatches, 3, 'gte', 3);
+test('gte checks for same type', valueDoesntMatch, 3, 'gte', '2');
+test('gte checks for same type', valueDoesntMatch, 3, 'gte', '3');
+test('gte checks for same type', keyMatches, 'abc', 'gte', 'ab');
+test('gte checks for same type', keyMatches, 'abc', 'gte', 'abc');
+test('gte checks for same type', keyDoesntMatch, 'abc', 'gte', 2 as any);
+
+test('lte checks for same type', valueMatches, 3, 'lte', 4);
+test('lte checks for same type', valueMatches, 3, 'lte', 3);
+test('lte checks for same type', valueDoesntMatch, 3, 'lte', '4');
+test('lte checks for same type', valueDoesntMatch, 3, 'lte', '3');
+test('lte checks for same type', keyMatches, 'abc', 'lte', 'abd');
+test('lte checks for same type', keyMatches, 'abc', 'lte', 'abc');
+test('lte checks for same type', keyDoesntMatch, 'abc', 'lte', 4 as any);
+
+type PredicateOp = 'exists' | 'isNull';
+const predicateHolds: Macro<[any, PredicateOp], Context> = async (t, testValue, operator) => {
+  for (const [path, value] of nest(testValue)) {
+    // TypeScript type system not strong enough to deduce this always
+    // creates a legal filter, so just cast.
+    const filter = { operator, path } as Filter;
+    t.true(await t.context.match({ key: 'abc', value }, filter));
+  }
+};
+predicateHolds.title = (providedTitle = '', testValue, operator) =>
+  `${providedTitle} ${operator}(${JSON.stringify(testValue)})`;
+
+const predicateFails: Macro<[any, PredicateOp], Context> = async (t, testValue, operator) => {
+  for (const [path, value] of nest(testValue)) {
+    // TypeScript type system not strong enough to deduce this always
+    // creates a legal filter, so just cast.
+    const filter = { operator, path } as Filter;
+    t.false(await t.context.match({ key: 'abc', value }, filter));
+  }
+};
+predicateFails.title = (providedTitle = '', testValue, operator) =>
+  `${providedTitle} !${operator}(${JSON.stringify(testValue)})`;
+
+test('exists checks for undefined', predicateHolds, null, 'exists');
+test('exists checks for undefined', predicateHolds, 0, 'exists');
+test('exists checks for undefined', predicateHolds, 'a', 'exists');
+test('exists sanity', async (t) => {
+  for (const path of [['value', 'a']]) {
+    const filter = { operator: 'exists' as 'exists', path };
+    t.false(await t.context.match({ key: 'abc', value: {} }, filter));
+  }
+  for (const path of [['key'], ['value']]) {
+    const filter = { operator: 'exists' as 'exists', path };
+    t.true(await t.context.match({ key: 'abc', value: {} }, filter));
+  }
+});
+
+test('isNull checks for null', predicateHolds, null, 'isNull');
+test('isNull checks for null', predicateFails, 0, 'isNull');
```

Comment:
somehow `test('isNull checks for null', predicateFails, undefined, 'isNull');` got dropped here during the move

Reply:
Changed the tests up, you can't create a document with an `undefined` value.